### PR TITLE
Add the ability to set trust managers for Apache, URLConnection, Netty

### PR DESCRIPTION
**Pull request by abrooksv**
## Description
Ability to set a custom trust manager for Apache

## Motivation and Context
Allows seamless setting of certs in IntelliJ

## Testing
Manually, got any ideas? 😄 

## Screenshots (if appropriate)
Screen Shot 2020-01-22 at 3 26 20 PM

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [ ] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

**Comment by abrooksv** [comment]
Refactored URLConnectionClient

1. Trust all no longer sets it JVM wide
2. You can now set key manager on it to bring it closer to Apache
3. You can now set trust manager on it to bring it in line with this change

Will look into testing tomorrow, as well as an optimization of not creating the SSL context each time in URLConnectionClient

**Comment by abrooksv** [comment]
I will also add it to Netty tomorrow for full coverage

**Comment by abrooksv** [comment]
Added Netty support, only create the SSL context once in UrlConnection.

Will look into adding tests into the SPI tests tomorrow using Wiremock and self signed certs

**Comment by abrooksv** [comment]
Tests added for sync clients

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/NettyConfiguration.java [hunk 1]
```diff
@@ -92,4 +89,12 @@ public boolean reapIdleConnections() {
     public TlsKeyManagersProvider tlsKeyManagersProvider() {
         return configuration.get(SdkHttpConfigurationOption.TLS_KEY_MANAGERS_PROVIDER);
     }
+
+    public TlsTrustManagersProvider tlsTrustManagersProvider() {
+        return configuration.get(SdkHttpConfigurationOption.TLS_TRUST_MANAGERS_PROVIDER);
+    }
+
+    public boolean trustAllCertificates() {
+        return configuration.get(TRUST_ALL_CERTIFICATES);
```
Comment (spfink) [inline http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/NettyConfiguration.java]: Not directly related to your change, but can you make static vs not static import consistent?

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java [hunk 2]
```diff
@@ -581,6 +599,10 @@ private HostnameVerifier getHostNameVerifier(AttributeMap standardOptions) {
 
         private SSLContext getSslContext(AttributeMap standardOptions) {
             TrustManager[] trustManagers = null;
+            if (standardOptions.containsKey(SdkHttpConfigurationOption.TLS_TRUST_MANAGERS_PROVIDER)) {
```
Comment (spfink) [inline http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java]: Add validation here (and other clients) that both this and trust all can't both be set?

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/NettyConfiguration.java (hunk 1 above)
Comment (abrooksv) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/NettyConfiguration.java]: Done

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java (hunk 2 above)
Comment (abrooksv) [reply http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java]: Done

**Comment by abrooksv** [comment]
Only thing left is figuring out netty tests

**Comment by abrooksv** [comment]
Latest revision adds in netty test

**Comment by abrooksv** [comment]
Ready for review

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClient.java [hunk 3]
```diff
@@ -82,9 +73,13 @@
 
     // Override connection idle timeout for Netty http client to reduce the frequency of "server failed to complete the
     // response error". see https://github.com/aws/aws-sdk-java-v2/issues/1122
+    // CHECKSTYLE:OFF - Lines are too long
     private static final AttributeMap NETTY_HTTP_DEFAULTS = AttributeMap.builder()
-                                                                        .put(CONNECTION_MAX_IDLE_TIMEOUT, Duration.ofSeconds(5))
+                                                                        .put(
+                                                                            SdkHttpConfigurationOption.CONNECTION_MAX_IDLE_TIMEOUT,
```
Comment (spfink) [inline http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClient.java]: This formatting is weird

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java [hunk 4]
```diff
@@ -580,14 +590,22 @@ private HostnameVerifier getHostNameVerifier(AttributeMap standardOptions) {
         }
 
         private SSLContext getSslContext(AttributeMap standardOptions) {
+            Validate.isTrue(standardOptions.get(SdkHttpConfigurationOption.TLS_TRUST_MANAGERS_PROVIDER) == null ||
+                            !standardOptions.get(SdkHttpConfigurationOption.TRUST_ALL_CERTIFICATES),
+                            "The tlsTrustManagersProvider and trust all can't both be configured.");
```
Comment (spfink) [inline http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java]: A TlsTrustManagerProvider can't be provided if TrustAllCertificates is also set

File: http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClient.java (hunk 3 above)
Comment (abrooksv) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClient.java]: Yaaa, normal format puts it past the line limit. Disable Checkstyle for this block?

Comment (spfink) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClient.java]: New line AttributeMap.builder()?

Comment (abrooksv) [reply http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClient.java]: Works for me!

**Pull request merged**